# **Challenge 2 Telecom X**

Foco do Projeto:

O principal objetivo deste desafio é desenvolver suas habilidades em ETL (Extract, Transform, Load) com Python. Os dados para este desafio estão disponíveis em uma API.


Etapas:
## - 1º Extração
## - 2º Tranformação
## - 3º Carga e Análise
## - 4º Relatório Final

# **1º Extração**

In [1]:
import pandas as pd
import requests

In [2]:
# url do dataset bruto

url = 'https://raw.githubusercontent.com/ingridcristh/challenge2-data-science/refs/heads/main/TelecomX_Data.json'

# 2º 🔧 **Transformação**

In [3]:
# Extraindo e já aplanando a estrutura JSON para o nosso df
df = pd.json_normalize(pd.read_json(url).to_dict(orient='records'))

# Renomeando as colunas para português (Normalização de nomes)
traducao = {
    'customerID': 'ID_Cliente',
    'Churn': 'Evasao_Churn',
    'customer.gender': 'Genero',
    'customer.SeniorCitizen': 'Idoso',
    'customer.Partner': 'Parceiro',
    'customer.Dependents': 'Dependentes',
    'customer.tenure': 'Meses_Contrato',
    'phone.PhoneService': 'Servico_Telefone',
    'phone.MultipleLines': 'Multiplas_Linhas',
    'internet.InternetService': 'Servico_Internet',
    'internet.OnlineSecurity': 'Seguranca_Online',
    'internet.OnlineBackup': 'Backup_Online',
    'internet.DeviceProtection': 'Protecao_Dispositivo',
    'internet.TechSupport': 'Suporte_Tecnico',
    'internet.StreamingTV': 'Streaming_TV',
    'internet.StreamingMovies': 'Streaming_Filmes',
    'account.Contract': 'Tipo_Contrato',
    'account.PaperlessBilling': 'Fatura_Digital',
    'account.PaymentMethod': 'Metodo_Pagamento',
    'account.Charges.Monthly': 'Valor_Mensal',
    'account.Charges.Total': 'Valor_Total'
}
df.rename(columns=traducao, inplace=True)

df

,ID_Cliente,Evasao_Churn,Genero,Idoso,Parceiro,Dependentes,Meses_Contrato,Servico_Telefone,Multiplas_Linhas,Servico_Internet,...,Backup_Online,Protecao_Dispositivo,Suporte_Tecnico,Streaming_TV,Streaming_Filmes,Tipo_Contrato,Fatura_Digital,Metodo_Pagamento,Valor_Mensal,Valor_Total
0,0002-ORFBO,No,Female,0,Yes,Yes,9,Yes,No,DSL,...,Yes,No,Yes,Yes,No,One year,Yes,Mailed check,65.60,593.3
1,0003-MKNFE,No,Male,0,No,No,9,Yes,Yes,DSL,...,No,No,No,No,Yes,Month-to-month,No,Mailed check,59.90,542.4
2,0004-TLHLJ,Yes,Male,0,No,No,4,Yes,No,Fiber optic,...,No,Yes,No,No,No,Month-to-month,Yes,Electronic check,73.90,280.85
3,0011-IGKFF,Yes,Male,1,Yes,No,13,Yes,No,Fiber optic,...,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,98.00,1237.85
4,0013-EXCHZ,Yes,Female,1,Yes,No,3,Yes,No,Fiber optic,...,No,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,83.90,267.4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7262,9987-LUTYD,No,Female,0,No,No,13,Yes,No,DSL,...,No,No,Yes,No,No,One year,No,Mailed check,55.15,742.9
7263,9992-RRAMN,Yes,Male,0,Yes,No,22,Yes,Yes,Fiber optic,...,No,No,No,No,Yes,Month-to-month,Yes,Electronic check,85.10,1873.7
7264,9992-UJOEL,No,Male,0,No,No,2,Yes,No,DSL,...,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,50.30,92.75
7265,9993-LHIEB,No,Male,0,Yes,Yes,67,Yes,No,DSL,...,No,Yes,Yes,No,Yes,Two year,No,Mailed check,67.85,4627.65


In [4]:
# Aplicando a renomeação direto no objeto
df.rename(columns=traducao, inplace=True)

# 3. Tratamento de Tipos (A "mágica" sem o coerce)
# Primeiro trocamos espaços vazios por '0', depois convertemos para número
df['Valor_Total'] = df['Valor_Total'].replace(' ', '0').astype(float)

# Verificando se os tipos e nomes estão no esquema
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   ID_Cliente            7267 non-null   object 
 1   Evasao_Churn          7267 non-null   object 
 2   Genero                7267 non-null   object 
 3   Idoso                 7267 non-null   int64  
 4   Parceiro              7267 non-null   object 
 5   Dependentes           7267 non-null   object 
 6   Meses_Contrato        7267 non-null   int64  
 7   Servico_Telefone      7267 non-null   object 
 8   Multiplas_Linhas      7267 non-null   object 
 9   Servico_Internet      7267 non-null   object 
 10  Seguranca_Online      7267 non-null   object 
 11  Backup_Online         7267 non-null   object 
 12  Protecao_Dispositivo  7267 non-null   object 
 13  Suporte_Tecnico       7267 non-null   object 
 14  Streaming_TV          7267 non-null   object 
 15  Streaming_Filmes     

In [5]:
# Verificando se há IDs duplicados

duplicados = df['ID_Cliente'].duplicated().sum()
print(f"Registros duplicados encontrados: {duplicados}")

Registros duplicados encontrados: 0


In [6]:
df.head()

,ID_Cliente,Evasao_Churn,Genero,Idoso,Parceiro,Dependentes,Meses_Contrato,Servico_Telefone,Multiplas_Linhas,Servico_Internet,...,Backup_Online,Protecao_Dispositivo,Suporte_Tecnico,Streaming_TV,Streaming_Filmes,Tipo_Contrato,Fatura_Digital,Metodo_Pagamento,Valor_Mensal,Valor_Total
0,0002-ORFBO,No,Female,0,Yes,Yes,9,Yes,No,DSL,...,Yes,No,Yes,Yes,No,One year,Yes,Mailed check,65.6,593.30
1,0003-MKNFE,No,Male,0,No,No,9,Yes,Yes,DSL,...,No,No,No,No,Yes,Month-to-month,No,Mailed check,59.9,542.40
2,0004-TLHLJ,Yes,Male,0,No,No,4,Yes,No,Fiber optic,...,No,Yes,No,No,No,Month-to-month,Yes,Electronic check,73.9,280.85
3,0011-IGKFF,Yes,Male,1,Yes,No,13,Yes,No,Fiber optic,...,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,98.0,1237.85
4,0013-EXCHZ,Yes,Female,1,Yes,No,3,Yes,No,Fiber optic,...,No,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,83.9,267.40


# **3º 📊 Carga**

In [7]:
# Salvando o DataFrame em outra variável e guardando o csv

dados = df.copy()

dados.to_csv('dados_telecom_x_limpo.csv', index=False)

In [8]:
dados_limpos = pd.read_csv('/content/dados_telecom_x_limpo.csv')

dados_limpos

,ID_Cliente,Evasao_Churn,Genero,Idoso,Parceiro,Dependentes,Meses_Contrato,Servico_Telefone,Multiplas_Linhas,Servico_Internet,...,Backup_Online,Protecao_Dispositivo,Suporte_Tecnico,Streaming_TV,Streaming_Filmes,Tipo_Contrato,Fatura_Digital,Metodo_Pagamento,Valor_Mensal,Valor_Total
0,0002-ORFBO,No,Female,0,Yes,Yes,9,Yes,No,DSL,...,Yes,No,Yes,Yes,No,One year,Yes,Mailed check,65.60,593.30
1,0003-MKNFE,No,Male,0,No,No,9,Yes,Yes,DSL,...,No,No,No,No,Yes,Month-to-month,No,Mailed check,59.90,542.40
2,0004-TLHLJ,Yes,Male,0,No,No,4,Yes,No,Fiber optic,...,No,Yes,No,No,No,Month-to-month,Yes,Electronic check,73.90,280.85
3,0011-IGKFF,Yes,Male,1,Yes,No,13,Yes,No,Fiber optic,...,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,98.00,1237.85
4,0013-EXCHZ,Yes,Female,1,Yes,No,3,Yes,No,Fiber optic,...,No,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,83.90,267.40
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7262,9987-LUTYD,No,Female,0,No,No,13,Yes,No,DSL,...,No,No,Yes,No,No,One year,No,Mailed check,55.15,742.90
7263,9992-RRAMN,Yes,Male,0,Yes,No,22,Yes,Yes,Fiber optic,...,No,No,No,No,Yes,Month-to-month,Yes,Electronic check,85.10,1873.70
7264,9992-UJOEL,No,Male,0,No,No,2,Yes,No,DSL,...,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,50.30,92.75
7265,9993-LHIEB,No,Male,0,Yes,Yes,67,Yes,No,DSL,...,No,Yes,Yes,No,Yes,Two year,No,Mailed check,67.85,4627.65


# **4º 📊 Análise Exploratória**

In [10]:
# Calculando a Taxa Geral de Churn

taxa_evasao = dados_limpos['Evasao_Churn'].value_counts(normalize=True) *100
print(f'Taxa de Churn Atual: {taxa_evasao['Yes']:.2f}%')
print("-" * 30)
print(dados_limpos['Evasao_Churn'].value_counts())

Taxa de Churn Atual: 26.54%
------------------------------
Evasao_Churn
No     5174
Yes    1869
Name: count, dtype: int64


In [12]:
# Vamos tentar entender o perfil do cliente que evade

# Criando uma Pivot Table para ver a taxa de Churn por Contrato e Idoso
# Nota: Como o Churn é 'Yes'/'No', vamos converter temporariamente para 1/0 para tirar a média (taxa)
dados_limpos['Churn_Numerico'] = dados_limpos['Evasao_Churn'].map({'Yes': 1, 'No': 0})

perfil = dados_limpos.pivot_table(index='Tipo_Contrato', columns='Idoso', values='Churn_Numerico', aggfunc='mean')

# Aplicando Styler para formatar como porcentagem e gradiente verde
pivot_perfil.style.format("{:.2%}")\
            .background_gradient(cmap='Greens')

Idoso,0,1
Tipo_Contrato,,
Month-to-month,39.57%,54.65%
One year,10.68%,15.26%
Two year,2.71%,4.14%


## Insights:

- Taxa de 26,54% de Churn é um número alto para telecomunicações, isso significa que mais de 1/4 da base está indo embora. Temos um problema real para resolver.

- O Vilão Principal: O contrato Month-to-month (mês a mês). Mesmo para não-idosos (39.57%), a evasão é altíssima comparada aos contratos de dois anos (2.71%).

- Fator de Risco (Idosos): Em todos os tipos de contrato, a taxa de evasão dos idosos é consideravelmente maior.

- A Solução Térmica: O contrato de dois anos "congela" o Churn. A estratégia da empresa deveria ser converter clientes mensais em anuais/bienais o mais rápido possível.

In [13]:
# Analisando a média de Valor Mensal para quem sai e quem fica
analise_financeira = dados_limpos.groupby('Evasao_Churn')['Valor_Mensal'].agg(['mean', 'median', 'std'])

# Formatando para exibir como moeda
analise_financeira.style.format("R$ {:.2f}").background_gradient(cmap='YlOrRd')


,mean,median,std
Evasao_Churn,,,
No,R$ 61.27,R$ 64.43,R$ 31.09
Yes,R$ 74.44,R$ 79.65,R$ 24.67


# **5. 📄Relatorio Final**